### ISSUE TITLE
Matching Lua's "long strings"

### ISSUE THREAD

**Issue by jonas-schievink** [kevinmehall/rust-peg#68]
I'm trying to write a parser that matches Lua's long string syntax. Basically, "[" followed by any number of "=" (including 0) followed by "[" opens the string, and "]" followed by **the same number** of "=" followed by "]" closes it.

Examples of long strings:
* `[[simple long string]]`
* `[=["level 1" long string]]this is still part of the string]==]as is this]=]`

I didn't come up with a parser that works, it's probably not even possible. This is what I currently have:

```rust
mlstring_char -> char
    = [^\]] { match_str.char_at(0) }

mlstring_inner -> String
    = "=" s:mlstring_inner "="      { s }
    / "[" chars:mlstring_char* "]"  { chars.into_iter().collect() }

string -> String
    = "[" s:mlstring_inner "]" { s }
```

This obviously doesn't work: Long strings end with the first "]" inside of them. If PEG wasn't greedy, I could just match any char inside `mlstring_inner`, but rust-peg doesn't seem to have any non-greedy mode.
Another way to solve this would be to reference captured strings in the pattern (like some Regex implementations allow). That way, I could just capture `"="*` in the opening braces, and reference it in the closing brace. The match would only succeed if both strings are the same, ie. if both braces use the same number of "="s.
The third way I've come up with is the most flexible solution: Run arbitrary rust code to determine if a rule should match. This way, I could just compare the number of "="s, and match the end of the long string if it has an equal number of "="s.

Is this somehow possible without modifying rust-peg?

**Comment by untitaker** [comment]
For your examples, what would the parsed string look like? (i.e. are the inner brackets part of the content)

**Comment by jonas-schievink** [comment]
No, the inner brackets are "long brackets" (at least that's what Lua calls them) and delimit the string. So the long string `[===[a]===]` would just contain "a".

**Comment by untitaker** [comment]
Did you try replacing `mlstring_char` with a `.`?

**Comment by jonas-schievink** [comment]
Then you would have `.*` in there, which always matches until the end of the input (unless I misunderstood PEGs greedy behaviour).

I can write a grammar that matches until the next long bracket, regardless of its length, but matching against a long bracket of the appropriate length requires backreferences or "conditional matching", where a Rust expression is evaluated and can return false to make the match fail (so I can compare the number of "="s and let the match fail if they're not equal). I implemented backreferences in #69. They also allow other neat stuff, for example matching XML tags where the content of the opening tag equals that of the closing tag.

Here's a grammar that demonstrates this:

```rust
string_inner -> String
    = (!("]" "="* "]") .)* { match_str.to_string() }

#[pub]
string -> String
    = "[" "="* "[" s:string_inner "]" "="* "]" {s}
```

This would also match `[===[a]]` and return `a`, which I do not want, since the brackets have a different length.

**Comment by untitaker** [comment]
>Then you would have .* in there, which always matches until the end of the input (unless I misunderstood PEGs greedy behaviour).

rust-peg wouldn't be able to match the required closing brackets if `.*` consumes everything. It does have some kind of backtracking support, I'm not sure if it applies here. You'd have to try it out. If it works, it would be inefficient for sure.

**Comment by kevinmehall** [comment]
Is this resolved to your satisfaction by #71?

**Comment by jonas-schievink** [comment]
Yes, conditional matching will allow this. Thanks!
